File: java/patterns-use-cases/src/main/java/my/example/sagas/clients/CarRentalClient.java
package my.example.sagas.clients;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class CarRentalClient {
  private static final Logger logger = LogManager.getLogger(CarRentalClient.class);

  public record CarRequest(String pickupLocation, String rentalDate) {}

  public static void book(String customerId, CarRequest request) {
    // this should implement the communication with the rental
    // provider's APIs
    logger.info("Car rental reservation created for customer: {}", customerId);
  }

  public static void cancel(String bookingId) {
    // this should implement the communication with the rental
    // provider's APIs
    logger.info("Car rental reservation cancelled with id: {}", bookingId);
  }
}

File: kotlin/patterns-use-cases/src/main/kotlin/my/example/sagas/clients/CarRentalClient.kt
package my.example.sagas.clients

import kotlinx.serialization.Serializable
import org.apache.logging.log4j.LogManager

@Serializable
data class CarRentalBookingRequest(val pickupLocation: String, val rentalDate: String)

private val logger = LogManager.getLogger("CarRentals")

fun bookCar(customerId: String, request: CarRentalBookingRequest) {
  // this should implement the communication with the rental
  // provider's APIs
  logger.info("Car rental reservation created for customer: {}", customerId)
}

fun cancelCar(customerId: String) {
  // this should implement the communication with the rental
  // provider's APIs
  logger.info("Car rental reservation cancelled for customer: {}", customerId)
}
